spark/src/main/scala/org/apache/spark/sql/delta/files/DelayedCommitProtocol.scala

```diff
@@ -123,33 +126,49 @@ class DelayedCommitProtocol(
     }
   }
 
-  protected def parsePartitions(dir: String): Map[String, String] = {
-    // TODO: timezones?
+  protected def parsePartitions(dir: String,
+                                taskContext: TaskAttemptContext): Map[String, String] = {
     // TODO: enable validatePartitionColumns?
+    val utcTimestampPartitionValues = taskContext.getConfiguration.getBoolean(
+      DeltaSQLConf.UTC_TIMESTAMP_PARTITION_VALUES.key, true) &&
+      taskContext.isInstanceOf[PartitionedTaskAttemptContextImpl] &&
+      taskContext.asInstanceOf[PartitionedTaskAttemptContextImpl]
+        .partitionAttr.exists(attribute => attribute.dataType.isInstanceOf[TimestampType])
+    val sessionTimeZone = getZoneId(SQLConf.get.sessionLocalTimeZone)
+    val partitionCols: Map[String, DataType] = if (utcTimestampPartitionValues) {
+      taskContext.asInstanceOf[PartitionedTaskAttemptContextImpl]
+          .partitionAttr.groupBy(attr => attr.name)
```

Comment:
This is for user-defined data types for partition columns. To preserve existing behavior as much as possible (and therefore add as little risk as possible), we should only include timestamp-valued columns in this list. The code will use the existing inference logic for any other columns.

Reply:
Agreed, that minimizes the risk by preserving the existing behavior as much as possible